opnode/p2p/gossip.go

```diff
@@ -0,0 +1,427 @@
+const maxOutboundQueue = 256
+const maxValidateQueue = 256
+const globalValidateThrottle = 512
+
+// Message domains, the msg id function uncompresses to keep data monomorphic,
+// but invalid compressed data will need a unique different id.
+
+var MessageDomainInvalidSnappy = [4]byte{0, 0, 0, 0}
+var MessageDomainValidSnappy = [4]byte{1, 0, 0, 0}
+
+const MaxGossipSize = 1 << 20
+
+func blocksTopicV1(cfg *rollup.Config) string {
+	return fmt.Sprintf("/optimism/%s/0/blocks", cfg.L2ChainID.String())
+}
+
+// BuildSubscriptionFilter builds a simple subscription filter,
+// to help protect against peers spamming useless subscriptions.
+func BuildSubscriptionFilter(cfg *rollup.Config) pubsub.SubscriptionFilter {
+	return pubsub.NewAllowlistSubscriptionFilter(blocksTopicV1(cfg)) // add more topics here in the future, if any.
+}
+
+var msgBufPool = sync.Pool{New: func() any {
+	// note: the topic validator concurrency is limited, so pool won't blow up, even with large pre-allocation.
+	x := make([]byte, 0, MaxGossipSize)
+	return &x
+}}
+
+// BuildMsgIdFn builds a generic message ID function for gossipsub that can handle compressed payloads,
+// mirroring the eth2 p2p gossip spec.
+func BuildMsgIdFn(cfg *rollup.Config) pubsub.MsgIdFunction {
+	return func(pmsg *pb.Message) string {
+		valid := false
+		var data []byte
+		// If it's a valid compressed snappy data, then hash the uncompressed contents.
+		// The validator can throw away the message later when recognized as invalid,
+		// and the unique hash helps detect duplicates.
+		dLen, err := snappy.DecodedLen(pmsg.Data)
+		if err == nil && dLen <= MaxGossipSize {
+			res := msgBufPool.Get().(*[]byte)
+			defer msgBufPool.Put(res)
+			if data, err = snappy.Decode((*res)[:0], pmsg.Data); err == nil {
+				*res = data // if we ended up growing the slice capacity, fine, keep the larger one.
+				valid = true
+			}
+		}
+		if data == nil {
+			data = pmsg.Data
+		}
+		h := sha256.New()
+		if valid {
+			h.Write(MessageDomainValidSnappy[:])
+		} else {
+			h.Write(MessageDomainInvalidSnappy[:])
+		}
+		// The chain ID is part of the gossip topic, making the msg id unique
+		topic := pmsg.GetTopic()
+		var topicLen [8]byte
+		binary.LittleEndian.PutUint64(topicLen[:], uint64(len(topic)))
+		h.Write(topicLen[:])
+		h.Write([]byte(topic))
+		h.Write(data)
+		// the message ID is shortened to save space, a lot of these may be gossiped.
+		return string(h.Sum(nil)[:20])
+	}
+}
+
+func BuildGlobalGossipParams(cfg *rollup.Config) pubsub.GossipSubParams {
+	params := pubsub.DefaultGossipSubParams()
+	params.D = 8                                      // topic stable mesh target count
+	params.Dlo = 6                                    // topic stable mesh low watermark
+	params.Dhi = 12                                   // topic stable mesh high watermark
+	params.Dlazy = 6                                  // gossip target
+	params.HeartbeatInterval = 500 * time.Millisecond // frequency of heartbeat, seconds
```

Comment:
Doc comment does not match the actual interval.

Reply:
The doc comment is copy-paste from the spec. Spec says `0.5`, but Go uses `time.Duration` instead. Will fix the comment